File: 이진트리동적할당/tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Student {
	char name[100];
	int midtermScore;
	int finalScore;
	struct Student* left;
	struct Student* right;

} Student;


Student* createNode(char name[], int midtermScore, int finalScore) {
	Student* newNode = (Student*)malloc(sizeof(Student));
	strcpy(newNode->name, name);
	newNode->midtermScore = midtermScore;
	newNode->finalScore = finalScore;	
	newNode->left = NULL;
	newNode->right = NULL;
	return newNode;
}

Student* insert(Student* root, char name[], int midtermScore, int finalScore) {
	if (root == NULL) {
		return createNode(name, midtermScore, finalScore);
	}

	if (strcmp(name, root->name) < 0) {
		root->left = insert(root->left, name, midtermScore, finalScore);
	} else if (strcmp(name, root->name) > 0) {
		root->right = insert(root->right, name, midtermScore, finalScore);
	}

	return root;
}

void inorderTraversal(Student* root) {
	if (root != NULL) {
		inorderTraversal(root->left);
		printf("Name: %s, Midterm Score: %d, Final Score: %d\n", root->name, root->midtermScore, root->finalScore);
		inorderTraversal(root->right);
	}
}

void deallocateTree(Student* root) {
	if (root != NULL) {
		deallocateTree(root->left);
		deallocateTree(root->right);
		free(root);
	}
}

int main() {
	int numStudents;
	printf("Enter the number of students: ");
	scanf("%d", &numStudents);
	Student* root = NULL;

	for (int i = 0; i < numStudents; i++) {
		char name[100];
		int midtermScore, finalScore;
	
		printf("Enter name of student %d: ", i + 1);
		scanf("%s", name);

		printf("Enter midterm score for %s: ", name);
		scanf("%d", &midtermScore);

		printf("Enter final score for %s: ", name);
		scanf("%d", &finalScore);

		root = insert(root, name, midtermScore, finalScore);
	
	}


	printf("\nStudent Information (Inorder Traversal):\n");
	inorderTraversal(root);
	deallocateTree(root);

	return 0;
}

File: 이진트리동적할당/tree.rs
use std::collections::BTreeMap;
use std::io;

struct Student {  
    midterm_score: i32,
    final_score: i32,
}

fn main() {
    let mut num_students = String::new();
    println!("Enter the number of students: ");
    io::stdin().read_line(&mut num_students).expect("Failed to read line");
    let num_students: i32 = num_students.trim().parse().expect("Invalid input");
    let mut student_map = BTreeMap::new();

    
    for i in 0..num_students {
        let mut name = String::new();
        println!("Enter name of student {}: ", i + 1);
        io::stdin().read_line(&mut name).expect("Failed to read line");
        let name = name.trim().to_string();

        
        let mut midterm_score = String::new();
        println!("Enter midterm score for {}: ", name);
        io::stdin().read_line(&mut midterm_score).expect("Failed to read line");
        let midterm_score: i32 = midterm_score.trim().parse().expect("Invalid input");

        
        let mut final_score = String::new();
        println!("Enter final score for {}: ", name);
        io::stdin().read_line(&mut final_score).expect("Failed to read line");        
        let final_score: i32 = final_score.trim().parse().expect("Invalid input");
   
        student_map.insert(name, Student { midterm_score, final_score });
        
    }
    
    println!("\nStudent Information (Inorder Traversal):");
    for (name, scores) in &student_map {
        println!("Name: {}, Midterm Score: {}, Final Score: {}", name, scores.midterm_score, scores.final_score);
        
    }
}

